spec/system/consumer/split_checkout_spec.rb

```diff
@@ -207,10 +207,111 @@
         end
       end
 
-      describe "filling out delivery details" do
+      describe "filling out order details" do
         before do
           fill_out_details
           fill_out_billing_address
+          order.update(user_id: user.id)
+        end
+
+        context "billing address" do
+          describe "checking the default address box" do
+            before do
+              check "order_save_bill_address"
+              choose free_shipping.name
+            end
+
+            it "before saving has no default address" do
+              expect(user.bill_address).to be_nil
+              expect(order.bill_address).to be_nil
+            end
+
+            context "proceeding to payment" do
+              before do
+                proceed_to_payment
```

Comment:
In my opinion, `before` blocks are to set the scene and not to execute the code that is tested. Sometimes including this in a before-block can reduce duplication but I often find later that I want to add a test case that needs something else to happen first. And then I don't have the flexibility.

I would remove this `context` and include `proceed_to_payment` in the `it` block. That's less nesting, too. Well, and I just gave an example how that can work with the `change` matcher.

Reply:
> In my opinion, before blocks are to set the scene

Agree :+1: 

> and not to execute the code that is tested.

So far, I've had the impression that ideally the code to be tested should be ran on the `before` block, so that the `it` block contains only the assertions, and does not execute code.

While this is [not really explicitly stated here](https://relishapp.com/rspec/rspec-core/v/3-9/docs/hooks/before-and-after-hooks), it did gave me that impression. 

So, I've been aiming at adding one `before` block per context; Maybe more lines of code, but maybe more readable and maybe easier to spot when something fails? What do you think?

(Side note, and more generally: I've spotted lengthy `it` blocks with different code being ran and multiple assertions on our test suite. So an idea could be using the `before` and `after` blocks to run all the necessary code, but I have not yet tried this yet.)

> I would remove this context and include proceed_to_payment in the it block. 

In any case, I totally agree that in this specific case :+1: as the code to be tested is too interlinked with the assertion - because of the `change` helper - so here it makes total sense to remove it from the `before` block. I'll make that change :+1

Reply:
Interesting. When I read those docs, I see that the subject is initialised with `@thing = Thing.new` but then there are operations in the `it` block like adding to a collection.

I've also seen guides which recommend to structure the `it` block in three parts with setup, execution and verification:

```rb
it "does something" do
  setup1
  setup2

  code_execution

  expect(...)
  expect(...)
end
```

Anyway, maybe this just depends on the individual case. Sometimes, an example needs to do something before the code execution. So if you needed to change the customer before you `proceed_to_checkout` then you can't nest that in the same block that has the checkout already in the before block.

Other times, you want to check several assertions for one operation. Then you can put the operation in the before block or even in the `subject` and then you have each case with `it` just containing expects.

Reply:
> Anyway, maybe this just depends on the individual case.

Sounds good :+1:

Great that we've had this exchange. I often have such questions (style, strategy, etc.) so these discussion snippets are very helpful.